pkg/asset/ignition/node.go

```diff
@@ -0,0 +1,97 @@
+	"bytes"
+	"fmt"
+	"html/template"
+
+	"github.com/apparentlymart/go-cidr/cidr"
+	"github.com/ghodss/yaml"
+
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/ignition/templates"
+	"github.com/openshift/installer/pkg/ignition"
+	"github.com/openshift/installer/pkg/types"
+)
+
+const (
+	// keyPaiAssetPrivateKeyIndex is the index of the private key in a key-pair asset.
+	keyPairAssetPrivateKeyIndex = 0
+	// keyPaiAssetPublicKeyIndex is the index of the public key in a key-pair asset.
+	keyPairAssetPublicKeyIndex = 1
+)
+
+// addNodeSections add the ignition config sections that are common to both master
+// and worker nodes.
+func addNodeSections(ignitionConfig *ignition.Config, rootCA *asset.State) {
+	addCommonSystemd(ignitionConfig)
+	ignitionConfig.AddFiles(
+		ignition.NewFile("/etc/hosts", templates.EtcHostsFileContents),
+		ignition.NewFile("/etc/sysctl.d/10-max-user-watches.conf", templates.MaxUserWatchesFileContents),
+		ignition.NewFile("/etc/coreos/update.conf", templates.UpdateFileContents),
+		newIgnitionFileFromAsset("/etc/docker/certs.d/docker-registry.default.svc:5000/ca.crt", rootCA, keyPairAssetPublicKeyIndex),
+	)
+}
+
+// newIgnitionFileFromAsset creates an ignition-config file with the contents
+// from the specified index in the specified asset state.
+func newIgnitionFileFromAsset(path string, assetState *asset.State, contentIndex int) ignition.File {
+	contents := string(assetState.Contents[contentIndex].Data)
+	return ignition.NewFile(path, contents)
+}
+
+// getInstallConfigFromAsset unmarshals the install config stored in the specified
+// asset state.
+func getInstallConfigFromAsset(installConfigAssetState *asset.State) (*types.InstallConfig, error) {
+	var installConfig types.InstallConfig
+	if err := yaml.Unmarshal(installConfigAssetState.Contents[0].Data, &installConfig); err != nil {
+		return nil, err
+	}
+	return &installConfig, nil
+}
+
+// getClusterDNSIP gets the DNS IP for the cluster described by the specified
+// install config.
+func getClusterDNSIP(installConfig *types.InstallConfig) (string, error) {
+	ip, err := cidr.Host(&installConfig.ServiceCIDR, 10)
+	if err != nil {
+		return "", err
+	}
+
+	return ip.String(), nil
+}
+
+// getCloudProvider gets the cloud provider for the cluster described by the
+// specified install config.
+func getCloudProvider(installConfig *types.InstallConfig) string {
+	if installConfig.AWS != nil {
+		return "aws"
+	}
+	return ""
+}
+
+// getCloudProviderConfig gets the cloud provider config for the cluster
+// described by the specified install config.
+func getCloudProviderConfig(installConfig *types.InstallConfig) string {
+	return ""
+}
+
+// etcdName gets the etcd hostname for the specified master in the cluster
+// described by the specified install config.
+func etcdName(installConfig *types.InstallConfig, masterIndex int) string {
+	return fmt.Sprintf("%s-etcd-%d.%s", installConfig.Name, masterIndex, installConfig.BaseDomain)
```

Comment:
This is potentially going away with https://github.com/openshift/installer/pull/187.